### DISCUSSION TITLE
Seeking feedback on new footer design!

### DISCUSSION THREAD

**Discussion by pradyunsg** [pradyunsg/furo#374] [Ideas]
In 29d65fdb8c0d9ab760c6c50771975e44b4978a39, I've implemented a new footer design.

<img width="775" alt="Screenshot 2022-02-13 at 17 52 32" src="https://user-images.githubusercontent.com/3275593/153767896-a3b69813-a222-4bec-b56d-9fc26a2fc8c2.png">

I've felt for a while that one of the most unpolished portions of Furo was the footer at the bottom of the page. It was very much put together in a "eh, let's just get over the line" mode when I first wrote this theme. This new design contains the following notable changes:

- Smaller font size, to de-emphasise the content in the footer.
- The ability to remove the link to Sphinx and the @pradyunsg, by passing `html_show_sphinx = False`. It is not possible to remove the link to Furo.
- Footer icons! This is something that I really want feedback on: https://pradyunsg.me/furo/customisation/footer/
- Drop the somewhat confusing `show source` link, which as added to be "compliant" with `html_show_source`. That option is now ignored.

**Comment by francois-rozet** [comment]
I would like the ability to remove your personal website but keep the Sphinx and Furo links.

Love the footer icons. Is the current API compatible with popular sets like Font Awesome's icons? Would be a huge selling point.

**Reply by pradyunsg** [reply]
> Is the current API compatible with popular sets like Font Awesome's icons?

It should be, if you add font awesome to the fonts on the page (eg: via sphinx-design). Someone might want to try things out!

> I would like the ability to remove your personal website but keep the Sphinx and Furo links.

I understand, but I'm not too keen on that; both because I'm wary of adding too many knobs and because I'm not sure how I feel about letting people drop the credit. Coupling with Sphinx feels like the right tradeoff, for documentation authors who want brevity in the footer.

Now is not a good time for an extended discussion on this specific topic though, because all that'll serve to do is make me less interested in tweaking the footer further.

**Reply by pradyunsg** [reply]
>  Is the current API compatible with popular sets like Font Awesome's icons? Would be a huge selling point.

See https://pradyunsg.me/furo/customisation/footer/#font-awesome :)

**Comment by Jackenmen** [comment]
As I've said in #114, I think that even with the addition of the "Edit on GitHub" button, the "show source" link still has a purpose. Could you consider bringing it back and enhancing it with GitHub links (i.e. `https://furo.readthedocs.io/en/latest/customisation/footer.html` would link to `https://github.com/pradyunsg/furo/blob/main/docs/customisation/footer.md?plain=1`) rather than just the "raw" source links (linking to the source hosted in `_sources`)?

I see that you consider "show source" to be confusing, is it because of the name? It could be changed to "show source of this document" though I'm not sure if that's not too long for the footer. If that doesn't seem right, it could instead be a footer icon if there's an icon that would make it clear what it means (`<>` might be clear enough?). Right now the custom footer icons are using fixed URLs so a dynamic "show source" link is only easily possible if done natively by the theme, I think?

I'm actually not sure if there's a need for two separate buttons, perhaps the user should just be able to configure between "Edit on GitHub" and "Show Source on GitHub" behavior, with a meaningful icon for each? This is kind of how [sphinx-rtd-theme does it](https://sphinx-rtd-theme.readthedocs.io/en/stable/configuring.html#confval-vcs_pageview_mode), though they always show it as "Edit on GitHub", regardless of the chosen view mode. I think `?plain=1` as part of the URL when the blob is chosen would be nice but that's a relatively new feature which might be why this was never brought up to its authors. Bearing in mind that the default they chose is `blob` rather than `edit`, I think it might at least be worth considering 👀 

Btw, [Python docs also have a "Show Source" link that goes to the blob on GitHub](https://docs.python.org/3/library/functions.html).

**Reply by pradyunsg** [reply]
> I see that you consider "show source" to be confusing, is it because of the name?

No. That was added for being compliant with https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_copy_source, and I don't want to add something with the same wording that behaves differently.

Update: eh, it says sidebar, so I guess we weren't ever 100% compliant.

> Btw, Python docs also have a "Show Source" link that goes to the blob on GitHub.

At the cost of sounding smug, I know. :)

That's a custom template tho, that's fully specific to CPython: https://github.com/python/cpython/blob/3.10/Doc/tools/templates/customsourcelink.html

> Right now the custom footer icons are using fixed URLs so a dynamic "show source" link is only easily possible if done natively by the theme, I think?

That's correct. I have an implementation of this, but I've not added that in because I'm not a 100% on the icon to use. `<>` is fine -- the tricky bit is finding something that fits in with GitHub and RTD logos.

**Reply by francois-rozet** [reply]
> Could you consider bringing it back and enhancing it with GitHub links [...] ?

The problem with Show source is that it links, 90% of the time, to an autosummary/autodoc generated file which is completely irrelevant. Unless we can filter for which files the Show source button appears, it will mostly be confusing.

Also, for these generated files, one will not be able to link to GitHub because they do not even exist on the repo. Sphinx can do that because it copies all the source files, even generated.

---

By the way speaking of links to GitHub, I would rather have the [source] buttons in the signatures linking to the code on GitHub. It is possible with `sphinx.ext.linkcode`, but this creates a `reference external` instead of `reference internal`, which is not styled by the theme.

Also the [source] button often overlaps with the signature content. Tried to fix that with CSS, but I did not succeed because the [source] button comes before the permalink button ¶. In the end I had to write a `sed` regex to swap these two buttons... This likely is a Sphinx problem, not a Furo one, but I am not experienced enough to know how to fix that.

Here is a [link](https://github.com/francois-rozet/piqa/tree/master/sphinx) to my generation files. The relevant ones are `build.sh` where the `sed` commands are, and `static/custom.css` where I tweak the css to prevent an overlap. Note that I also replaced the [source] text by a `</>` icon.

**Reply by pradyunsg** [reply]
> It is possible with `sphinx.ext.linkcode`, but this creates a `reference external` instead of `reference internal`, which is not styled by the theme.

That's fair, please make a separate discussion topic for this. It's off-topic for this discussion since it's unrelated to the footer.

**Reply by pradyunsg** [reply]
> The problem with Show source is that it links, 90% of the time, to an autosummary/autodoc generated file which is completely irrelevant.

That's what the default Sphinx themes do, which is what this modelled. I've literally removed it based on the feedback that it's not useful, so I'm not even sure what point you're trying to make anymore.

**Reply by francois-rozet** [reply]
It was an answer to Jack. The main issue is that you will not be able to link to GitHub because the file does not even exist. Sphinx can do that because it copies all the source files, even generated. But on the repo, there is no generated file.

**Reply by pradyunsg** [reply]
Ah, in that case, I misunderstood what you were saying. 😅 

As a somewhat nit-picky/annoying suggestion: Consider quoting the thing you're responding to, when it's in a thread like this.

**Reply by pradyunsg** [reply]
> That's fair, please make a separate discussion topic for this. It's off-topic for this discussion since it's unrelated to the footer.

Well, don't bother -- it's fixed and will be in the next release https://github.com/pradyunsg/furo/commit/32b2fe5709c1d30cef981f936ce9592d0758bec0

**Reply by francois-rozet** [reply]
> Consider quoting the thing you're responding to

Fixed 👍

> it's fixed and will be in the next release

Thanks!

**Reply by Jackenmen** [reply]
> > Could you consider bringing it back and enhancing it with GitHub links [...] ?
> 
> The problem with Show source is that it links, 90% of the time, to an autosummary/autodoc generated file which is completely irrelevant. Unless we can filter for which files the Show source button appears, it will mostly be confusing.

There could be a custom function callback like there is with Sphinx's linkcode but that on the other hand makes it more complicated to implement both for Pradyun *and* for users so yeah, I see your point. I didn't think of it asI don't personally use autosummary partially because I find it lacking and partially because I didn't have that many use for it so far. I imagine quite a few don't use it since it's not *that* easy with all the shenanigans of auto-generated files.

**Comment by jdknight** [comment]
Just some generic comments, I'm good either way:

1) The copyright format on this extension is slightly different, in terms of ordering, than Spinx's base theme. Sphinx's base theme [defines the copyright symbol before the text statement](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/themes/basic/layout.html#L200), whereas the Furo theme switches that around. I do not mind either way, I just wanted to mention the difference.
2) The period in the second copyright section is not required. Sphinx's theme only applies a period entry to help serve as a separator between other sections (if ever used; personally, I wish Sphinx's base theme would opt to not inject the period character when not needed, but that is outside this scope). Since the statement is encloded on its own `div`, I feel that the period character should be dropped (allowing users who do not want a period to not have it by default; and users who want a period can just append it in their copyright define).
```
<a href="{{ path }}">Copyright</a> &#169; {{ copyright }}.
```
3) I feel that it would be more "proper" to exclude the Furo branding when `html_show_sphinx` is set to `False` (although this doesn't need to be discussed anymore than it already has in the other comment; just giving another opinion).
4) It would be nice in the case where page-specific source links are added (as mentioned in another comment about View/Edit links to GitHub) that it is easy enough to support other popular VCS services, as well as custom URLs. For a Confluence extension, we do [provide a way](https://sphinxcontrib-confluencebuilder.readthedocs.io/configuration/#confval-confluence_sourcelink) for users to configure a generated source link using their own customizable format string. Instead of having to deal with multiple HTML context entries, one specific entry could be used to handle this. Although, if the goal is just to provide users flexibility via the `footer_icons` dictionary, maybe Furo can pass in some key-value arguments into a `format` call on the `url` field before generating a footer for a given page.

**Reply by pradyunsg** [reply]
> The copyright format on this extension is slightly different, in terms of ordering, than Spinx's base theme.

This is intentional, and feels more in line with how the rest of the internet handles copyright declarations. Plus, Furo also has some logic to give you complete control if you include the word Copyright in your conf.py's `copyright`.

> The period in the second copyright section is not required. 

https://github.com/pradyunsg/furo/commit/c1140777389462eb883211d175d74dd3b36a18d8

> I feel that it would be more "proper" to exclude the Furo branding

Cool, and I don't. :)

> It would be nice in the case where page-specific source links are added (as mentioned in another comment about View/Edit links to GitHub) that it is easy enough to support other popular VCS services, as well as custom URLs.

Hmm... I view this footer as being a site footer rather than a page footer though; which is why I don't want to add page-specific things there.

**Comment by brobr** [comment]
Thanks for the theme; the icons in the footer are very helpful; with the possibility of svg-tags other icons with links can be made, which is fab. For example, the change that can be made when moving a project to Codeberg: https://codeberg.org/Codeberg/Community/issues/976

**Comment by mluerig** [comment]
I like the new design. One comment / question: not sure if I missed it, but is it at all possible (without fiddling with the html template) to add your name to the copyright, and a hyperlink that links to my personal page?

**Comment by carlosgmartin** [comment]
@pradyunsg Is it possible to make the footer icons bigger? Currently, the height of the container is quite small.

This can be seen on the [documentation website](https://pradyunsg.me/furo/customisation/logo/) with the small GitHub logo.
